Cannot build transmission 4.0.0-beta.2 in freebsd: duplicate explicit instantiation of ‘IntType tr_rand_int(IntType) [with IntType = unsigned int]’

### What is the issue?

I followed the steps in the wiki to build it first from Git and when that failed, I tried building it from the source which also failed with the same error

https://github.com/transmission/transmission/issues/4026 was raised already with the same error but the solution proposed there didn't work for me and since it was closed, I decided to open a new thread on that.

```
root@Transmission:/ # cd ~/Transmission/
root@Transmission:~/Transmission # cd build
root@Transmission:~/Transmission/build # make clean
root@Transmission:~/Transmission/build # git pull --rebase --prune
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 18 (delta 12), reused 17 (delta 12), pack-reused 0
Unpacking objects: 100% (18/18), 8.69 KiB | 306.00 KiB/s, done.
From https://github.com/transmission/transmission
   4549cc74a..e32cac98c  main             -> origin/main
 * [new branch]          fixup/peer-count -> origin/fixup/peer-count
Updating 4549cc74a..e32cac98c
Fast-forward
 macosx/CMakeLists.txt             | 6 ++++++
 macosx/GroupPopUpButtonCell.h     | 2 +-
 macosx/GroupPopUpButtonCell.mm    | 2 +-
 macosx/PriorityPopUpButtonCell.h  | 2 +-
 macosx/PriorityPopUpButtonCell.mm | 2 +-
 5 files changed, 10 insertions(+), 4 deletions(-)
root@Transmission:~/Transmission/build # git submodule update --recursive
root@Transmission:~/Transmission/build # cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo .. -DCMAKE_C_SYSTEM_INCLUDE_DIRECTORIES=/usr/local/include -DCMAKE_CXX_SYSTEM_INCLUDE_DIRECTORIES=/usr/local/include
-- Looking for clang-tidy
-- Looking for clang-tidy - not found
-- Configuring done
-- Generating done
-- Build files have been written to: /root/Transmission/build
root@Transmission:~/Transmission/build # make
[  0%] Creating directories for 'deflate'
[  1%] Performing download step (DIR copy) for 'deflate'
[  2%] No update step for 'deflate'
[  3%] No patch step for 'deflate'
[  3%] Performing configure step for 'deflate'
-- Configuring done
-- Generating done
-- Build files have been written to: /root/Transmission/build/third-party/deflate/src/deflate-build
[  4%] Performing build step for 'deflate'
Consolidate compiler generated dependencies of target deflate
[  8%] Building C object CMakeFiles/deflate.dir/lib/deflate_decompress.c.o
[ 16%] Building C object CMakeFiles/deflate.dir/lib/utils.c.o
[ 25%] Building C object CMakeFiles/deflate.dir/lib/x86/cpu_features.c.o
[ 33%] Building C object CMakeFiles/deflate.dir/lib/zlib_compress.c.o
[ 41%] Building C object CMakeFiles/deflate.dir/lib/arm/cpu_features.c.o
[ 50%] Building C object CMakeFiles/deflate.dir/lib/deflate_compress.c.o
[ 58%] Building C object CMakeFiles/deflate.dir/lib/adler32.c.o
[ 66%] Building C object CMakeFiles/deflate.dir/lib/zlib_decompress.c.o
[ 75%] Building C object CMakeFiles/deflate.dir/lib/crc32.c.o
[ 83%] Building C object CMakeFiles/deflate.dir/lib/gzip_decompress.c.o
[ 91%] Building C object CMakeFiles/deflate.dir/lib/gzip_compress.c.o
[100%] Linking C static library libdeflate.a
[100%] Built target deflate
[  5%] Performing install step for 'deflate'
Consolidate compiler generated dependencies of target deflate
[100%] Built target deflate
Install the project...
-- Install configuration: "RelWithDebInfo"
-- Installing: /root/Transmission/build/third-party/deflate/lib/libdeflate.a
-- Installing: /root/Transmission/build/third-party/deflate/include/libdeflate.h
-- Up-to-date: /root/Transmission/build/third-party/deflate/lib/cmake/deflate/deflate-config.cmake
-- Up-to-date: /root/Transmission/build/third-party/deflate/lib/cmake/deflate/deflate-config-relwithdebinfo.cmake
[  5%] Completed 'deflate'
[  5%] Built target deflate
[  6%] Creating directories for 'utp'
[  7%] Performing download step (DIR copy) for 'utp'
[  7%] No update step for 'utp'
[  8%] No patch step for 'utp'
[  9%] Performing configure step for 'utp'
-- Configuring done
-- Generating done
-- Build files have been written to: /root/Transmission/build/third-party/utp/src/utp-build
[  9%] Performing build step for 'utp'
Consolidate compiler generated dependencies of target libutp
[ 14%] Building CXX object CMakeFiles/libutp.dir/utp_api.cpp.o
[ 28%] Building CXX object CMakeFiles/libutp.dir/utp_callbacks.cpp.o
[ 42%] Building CXX object CMakeFiles/libutp.dir/utp_hash.cpp.o
[ 57%] Building CXX object CMakeFiles/libutp.dir/utp_internal.cpp.o
[ 71%] Building CXX object CMakeFiles/libutp.dir/utp_packedsockaddr.cpp.o
[ 85%] Building CXX object CMakeFiles/libutp.dir/utp_utils.cpp.o
[100%] Linking CXX static library libutp.a
[100%] Built target libutp
[  9%] Performing install step for 'utp'
Consolidate compiler generated dependencies of target libutp
[100%] Built target libutp
Install the project...
-- Install configuration: "RelWithDebInfo"
-- Installing: /root/Transmission/build/third-party/utp/lib/libutp.a
-- Up-to-date: /root/Transmission/build/third-party/utp/include
-- Up-to-date: /root/Transmission/build/third-party/utp/include/libutp
-- Installing: /root/Transmission/build/third-party/utp/include/libutp/utp_types.h
-- Installing: /root/Transmission/build/third-party/utp/include/libutp/utp.h
-- Up-to-date: /root/Transmission/build/third-party/utp/lib/cmake/libutp/libutp-targets.cmake
-- Up-to-date: /root/Transmission/build/third-party/utp/lib/cmake/libutp/libutp-targets-relwithdebinfo.cmake
-- Up-to-date: /root/Transmission/build/third-party/utp/lib/cmake/libutp/libutp-config.cmake
-- Up-to-date: /root/Transmission/build/third-party/utp/lib/cmake/libutp/libutp-config-version.cmake
-- Installing: /root/Transmission/build/third-party/utp/share/doc/libutp/LICENSE
-- Installing: /root/Transmission/build/third-party/utp/share/doc/libutp/README.md
[ 10%] Completed 'utp'
[ 10%] Built target utp
[ 11%] Building C object third-party/jsonsl/CMakeFiles/jsonsl.dir/jsonsl.c.o
[ 12%] Linking C static library libjsonsl.a
[ 12%] Built target jsonsl
[ 13%] Building C object third-party/wildmat/CMakeFiles/wildmat.dir/wildmat.c.o
[ 14%] Linking C static library libwildmat.a
[ 14%] Built target wildmat
[ 14%] Building CXX object tests/gtest/CMakeFiles/gtestall.dir/__/__/third-party/googletest/googletest/src/gtest-all.cc.o
[ 15%] Building CXX object tests/gtest/CMakeFiles/gtestall.dir/__/__/third-party/googletest/googletest/src/gtest_main.cc.o
[ 15%] Linking CXX static library libgtestall.a
[ 15%] Built target gtestall
[ 16%] Building CXX object libtransmission/CMakeFiles/transmission.dir/announce-list.cc.o
[ 16%] Building CXX object libtransmission/CMakeFiles/transmission.dir/announcer-http.cc.o
[ 17%] Building CXX object libtransmission/CMakeFiles/transmission.dir/announcer-udp.cc.o
[ 18%] Building CXX object libtransmission/CMakeFiles/transmission.dir/announcer.cc.o
[ 18%] Building CXX object libtransmission/CMakeFiles/transmission.dir/bandwidth.cc.o
[ 19%] Building CXX object libtransmission/CMakeFiles/transmission.dir/bitfield.cc.o
[ 19%] Building CXX object libtransmission/CMakeFiles/transmission.dir/block-info.cc.o
[ 20%] Building CXX object libtransmission/CMakeFiles/transmission.dir/blocklist.cc.o
[ 21%] Building CXX object libtransmission/CMakeFiles/transmission.dir/cache.cc.o
[ 21%] Building CXX object libtransmission/CMakeFiles/transmission.dir/clients.cc.o
[ 22%] Building CXX object libtransmission/CMakeFiles/transmission.dir/completion.cc.o
[ 23%] Building CXX object libtransmission/CMakeFiles/transmission.dir/crypto-utils-openssl.cc.o
[ 23%] Building CXX object libtransmission/CMakeFiles/transmission.dir/crypto-utils.cc.o
[ 24%] Building CXX object libtransmission/CMakeFiles/transmission.dir/error.cc.o
[ 25%] Building CXX object libtransmission/CMakeFiles/transmission.dir/file-piece-map.cc.o
[ 25%] Building CXX object libtransmission/CMakeFiles/transmission.dir/file-posix.cc.o
[ 26%] Building CXX object libtransmission/CMakeFiles/transmission.dir/file.cc.o
[ 26%] Building CXX object libtransmission/CMakeFiles/transmission.dir/handshake.cc.o
In file included from /root/Transmission/libtransmission/handshake.cc:20:
/root/Transmission/libtransmission/handshake.h:257:24: warning: declaration requires an exit-time destructor [-Wexit-time-destructors]
    static inline auto dh_pool_mutex_ = std::mutex{};
                       ^
1 warning generated.
[ 27%] Building CXX object libtransmission/CMakeFiles/transmission.dir/inout.cc.o
[ 28%] Building CXX object libtransmission/CMakeFiles/transmission.dir/log.cc.o
/root/Transmission/libtransmission/log.cc:56:6: warning: declaration requires an exit-time destructor [-Wexit-time-destructors]
auto log_state = tr_log_state{};
     ^
1 warning generated.
[ 28%] Building CXX object libtransmission/CMakeFiles/transmission.dir/magnet-metainfo.cc.o
[ 29%] Building CXX object libtransmission/CMakeFiles/transmission.dir/makemeta.cc.o
[ 30%] Building CXX object libtransmission/CMakeFiles/transmission.dir/net.cc.o
[ 30%] Building CXX object libtransmission/CMakeFiles/transmission.dir/open-files.cc.o
[ 31%] Building CXX object libtransmission/CMakeFiles/transmission.dir/peer-io.cc.o
/root/Transmission/libtransmission/peer-io.cc:108:28: error: use of undeclared identifier 'utp_create_socket'
        auto* const sock = utp_create_socket(session->utp_context);
                           ^
/root/Transmission/libtransmission/peer-io.cc:158:9: error: use of undeclared identifier 'utp_set_userdata'
        utp_set_userdata(socket_.handle.utp, this);
        ^
/root/Transmission/libtransmission/peer-io.cc:670:62: error: use of undeclared identifier 'utp_error_code_names'
    tr_logAddTraceIo(this, fmt::format("utp_on_error -- {}", utp_error_code_names[errcode]));
                                                             ^
/root/Transmission/libtransmission/peer-io.cc:675:39: error: use of undeclared identifier 'utp_error_code_names'
        tr_error_set(&error, errcode, utp_error_code_names[errcode]);
                                      ^
/root/Transmission/libtransmission/peer-io.cc:686:33: error: use of undeclared identifier 'UTP_RCVBUF'
    utp_context_set_option(ctx, UTP_RCVBUF, RcvBuf);
                                ^
/root/Transmission/libtransmission/peer-io.cc:694:12: error: unknown type name 'utp_callback_arguments'
        [](utp_callback_arguments* args) -> uint64
           ^
/root/Transmission/libtransmission/peer-io.cc:694:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:693:9: error: use of undeclared identifier 'UTP_ON_READ'
        UTP_ON_READ,
        ^
/root/Transmission/libtransmission/peer-io.cc:694:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:708:12: error: unknown type name 'utp_callback_arguments'
        [](utp_callback_arguments* args) -> uint64
           ^
/root/Transmission/libtransmission/peer-io.cc:708:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:707:9: error: use of undeclared identifier 'UTP_GET_READ_BUFFER_SIZE'
        UTP_GET_READ_BUFFER_SIZE,
        ^
/root/Transmission/libtransmission/peer-io.cc:708:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:720:12: error: unknown type name 'utp_callback_arguments'
        [](utp_callback_arguments* args) -> uint64
           ^
/root/Transmission/libtransmission/peer-io.cc:720:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:719:9: error: use of undeclared identifier 'UTP_ON_ERROR'
        UTP_ON_ERROR,
        ^
/root/Transmission/libtransmission/peer-io.cc:720:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:732:12: error: unknown type name 'utp_callback_arguments'
        [](utp_callback_arguments* args) -> uint64
           ^
/root/Transmission/libtransmission/peer-io.cc:732:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:731:9: error: use of undeclared identifier 'UTP_ON_OVERHEAD_STATISTICS'
        UTP_ON_OVERHEAD_STATISTICS,
        ^
/root/Transmission/libtransmission/peer-io.cc:732:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:745:12: error: unknown type name 'utp_callback_arguments'
        [](utp_callback_arguments* args) -> uint64
           ^
/root/Transmission/libtransmission/peer-io.cc:745:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
/root/Transmission/libtransmission/peer-io.cc:744:9: error: use of undeclared identifier 'UTP_ON_STATE_CHANGE'
        UTP_ON_STATE_CHANGE,
        ^
/root/Transmission/libtransmission/peer-io.cc:745:36: warning: unused parameter 'args' [-Wunused-parameter]
        [](utp_callback_arguments* args) -> uint64
                                   ^
10 warnings and 15 errors generated.
*** Error code 1

Stop.
make[2]: stopped in /root/Transmission/build
*** Error code 1

Stop.
make[1]: stopped in /root/Transmission/build
*** Error code 1

Stop.
make: stopped in /root/Transmission/build
```

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.0-beta.2

The problem persists on arm7 (Raspberry Pi 4).  Building from main:

```
/home/karl/AUR/transmission-cli-git/src/transmission/libtransmission/crypto-utils.cc:47:23: error: duplicate explicit instantiation of 'tr_rand_int<unsigned int>' [clang-diagnostic-error]
template unsigned int tr_rand_int(unsigned int upper_bound);
                      ^
/home/karl/AUR/transmission-cli-git/src/transmission/libtransmission/crypto-utils.cc:46:17: note: previous explicit instantiation is here
template size_t tr_rand_int(size_t upper_bound);
                ^
```

```
% gcc --version
gcc (GCC) 12.1.0
% uname -svr
Linux 5.15.84-1-rpi-ARCH #1 SMP Mon Dec 19 13:37:50 MST 2022
```

The offending lines are
https://github.com/transmission/transmission/blob/84566d42f023266dd5d38e9c99bc0947775dcacb/libtransmission/crypto-utils.cc#L46-L47

I guess on arm7 `size_t` and `unsigned int` are the same type. Looks like #3196?

--- a/libtransmission/crypto-utils.cc
+++ b/libtransmission/crypto-utils.cc
@@ -31,23 +31,6 @@ using namespace std::literals;
 
 ///
 
-template<class T>
-[[nodiscard]] T tr_rand_int(T upper_bound)
-{
-    static_assert(!std::is_signed<T>());
-    TR_ASSERT(upper_bound > std::numeric_limits<T>::min());
-
-    using dist_type = std::uniform_int_distribution<T>;
-    thread_local auto rng = tr_urbg<T>{};
-    thread_local auto dist = dist_type{};
-    return dist(rng, typename dist_type::param_type(0, upper_bound - 1));
-}
-
-template uint32_t tr_rand_int(uint32_t upper_bound);
-template uint64_t tr_rand_int(uint64_t upper_bound);
-
-///
-
 namespace
 {
 namespace ssha1_impl

--- a/libtransmission/crypto-utils.cc
+++ b/libtransmission/crypto-utils.cc
@@ -43,8 +43,8 @@ template<class T>
     return dist(rng, typename dist_type::param_type(0, upper_bound - 1));
 }
 
-template size_t tr_rand_int(size_t upper_bound);
-template unsigned int tr_rand_int(unsigned int upper_bound);
+template uint32_t tr_rand_int(uint32_t upper_bound);
+template uint64_t tr_rand_int(uint64_t upper_bound);
 
 ///
 

--- a/libtransmission/crypto-utils.h
+++ b/libtransmission/crypto-utils.h
@@ -12,6 +12,7 @@
 #include <limits>
 #include <memory>
 #include <optional>
+#include <random> // for std::uniform_int_distribution<T>
 #include <string>
 #include <string_view>
 
@@ -86,12 +87,6 @@ tr_x509_cert_t tr_x509_cert_new(void const* der, size_t der_length);
  * @brief Free X509 certificate returned by @ref tr_x509_cert_new.
  */
 void tr_x509_cert_free(tr_x509_cert_t handle);
-
-/**
- * @brief Returns a random number in the range of [0...upper_bound).
- */
-template<class IntType>
-[[nodiscard]] IntType tr_rand_int(IntType upper_bound);
 
 /**
  * @brief Fill a buffer with random bytes.
@@ -211,6 +206,20 @@ public:
 private:
     tr_salt_shaker<T, N> buf_;
 };
+
+/**
+ * @brief Returns a random number in the range of [0...upper_bound).
+ */
+template<class T>
+[[nodiscard]] T tr_rand_int(T upper_bound)
+{
+    static_assert(!std::is_signed<T>());
+    using dist_type = std::uniform_int_distribution<T>;
+
+    thread_local auto rng = tr_urbg<T>{};
+    thread_local auto dist = dist_type{};
+    return dist(rng, typename dist_type::param_type(0, upper_bound - 1));
+}
 
 /** @} */
 

--- a/tests/libtransmission/crypto-test-ref.h
+++ b/tests/libtransmission/crypto-test-ref.h
@@ -18,7 +18,6 @@
 #define tr_rand_buffer_crypto tr_rand_buffer_crypto_
 #define tr_rand_buffer_std tr_rand_buffer_std_
 #define tr_rand_int tr_rand_int_
-#define tr_rand_int_weak tr_rand_int_weak_
 #define tr_rand_obj tr_rand_obj_
 #define tr_salt_shaker tr_salt_shaker_
 #define tr_sha1 tr_sha1_
@@ -54,7 +53,6 @@
 #undef tr_rand_buffer_crypto
 #undef tr_rand_buffer_std
 #undef tr_rand_int
-#undef tr_rand_int_weak
 #undef tr_rand_obj
 #undef tr_salt_shaker
 #undef tr_sha1
@@ -85,7 +83,6 @@
 #define tr_rand_buffer_crypto_ tr_rand_buffer_crypto
 #define tr_rand_buffer_std_ tr_rand_buffer_std
 #define tr_rand_int_ tr_rand_int
-#define tr_rand_int_weak_ tr_rand_int_weak
 #define tr_rand_obj_ tr_rand_obj
 #define tr_salt_shaker_ tr_salt_shaker
 #define tr_sha1_ tr_sha1

